dependency-analysis/main.go

```diff
@@ -0,0 +1,357 @@
+// Copyright 2023 OpenSSF Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+	"strings"
+
+	"github.com/google/go-github/v46/github"
+	"golang.org/x/oauth2"
+)
+
+func main() {
+	vulnerabilities := ""
+	result := ""
+
+	repo := os.Getenv("GITHUB_REPOSITORY")
+	commitSHA := os.Getenv("GITHUB_SHA")
+	token := os.Getenv("GITHUB_TOKEN")
+	pr := os.Getenv("GITHUB_PR_NUMBER")
+	ghUser := os.Getenv("GITHUB_ACTOR")
+	fileName := os.Getenv("SCORECARD_CHECKS")
+	if err := Validate(token, repo, commitSHA, pr); err != nil {
+		log.Fatal(err)
+	}
+	// convert pr to int
+	prInt, err := strconv.Atoi(pr)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ownerRepo := strings.Split(repo, "/")
+	if len(ownerRepo) != 2 {
+		log.Fatal("invalid repo")
+	}
+	owner := ownerRepo[0]
+	repo = ownerRepo[1]
+	checks, err := GetScorecardChecks(fileName)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	ts := oauth2.StaticTokenSource(&oauth2.Token{AccessToken: token})
+	tc := oauth2.NewClient(context.Background(), ts)
+	client := github.NewClient(tc)
+	defaultBranch, err := getDefaultBranch(owner, repo, client)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	diff, err := GetDependencyDiff(owner, repo, token, defaultBranch, commitSHA)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	m := make(map[string]DependencyDiff)
+	for _, dep := range diff { //nolint:gocritic
+		m[dep.SourceRepositoryURL] = dep
+	}
+
+	for k, i := range m { //nolint:gocritic
+		url := strings.TrimPrefix(k, "https://")
+		scorecard, err := GetScorecardResult(url)
+		if err != nil {
+			if len(i.Vulnerabilities) > 0 {
```

Comment:
Why are we treating Vulns as a special case? 

I worry our data is inaccurate, ie we're using a dependency HEAD whereas a user is installing a particular package version. There are other tools that will do a better job at this than Scorecard, like dependabot.

I'm thinking we may even remove the vuln results entirely because of how poor our data may be. Otherwise users will be confused and unhappy that we're claiming there are vulns in a package if there are none.

Reply:
OK. I agree. I can remove the Vulns.